drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,952 @@
+	LL_DMA_PDATAALIGN_BYTE,
+	LL_DMA_PDATAALIGN_HALFWORD,
+	LL_DMA_PDATAALIGN_WORD,
+};
+
+struct dma_stm32_stream {
+	u32_t direction;
+	bool busy;
+	void *callback_arg;
+	void (*dma_callback)(void *arg, u32_t id,
+			     int error_code);
+};
+
+struct dma_stm32_data {
+	int max_streams;
+	struct dma_stm32_stream *streams;
+};
+
+struct dma_stm32_config {
+	struct stm32_pclken pclken;
+	void (*config_irq)(struct device *dev);
+
+	bool support_m2m;
+	u32_t base;
+};
+
+/* DMA burst length */
+#define BURST_TRANS_LENGTH_1			0
+
+/* Maximum data sent in single transfer (Bytes) */
+#define DMA_STM32_MAX_DATA_ITEMS		0xffff
+
+static void dma_stm32_dump_stream_irq(struct device *dev, u32_t id)
+{
+	const struct dma_stm32_config *config = dev->config->config_info;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	LOG_INF("tc: %d, ht: %d, te: %d, dme: %d, fe: %d",
+		 func_is_active_tc[id](dma),
+		 func_is_active_ht[id](dma),
+		 func_is_active_te[id](dma),
+		 func_is_active_dme[id](dma),
+		 func_is_active_fe[id](dma));
+#else
+	LOG_INF("tc: %d, ht: %d, te: %d, gi: %d",
+		 func_is_active_tc[id](dma),
+		 func_is_active_ht[id](dma),
+		 func_is_active_te[id](dma),
+		 func_is_active_gi[id](dma));
+#endif
+}
+
+static void dma_stm32_clear_stream_irq(struct device *dev, u32_t id)
+{
+	const struct dma_stm32_config *config = dev->config->config_info;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+
+	func_clear_tc[id](dma);
+	func_clear_ht[id](dma);
+	func_clear_te[id](dma);
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	func_clear_dme[id](dma);
+	func_clear_fe[id](dma);
+#else
+	func_clear_gi[id](dma);
+#endif
+}
+
+static void dma_stm32_irq_handler(void *arg)
+{
+	struct device *dev = arg;
+	struct dma_stm32_data *data = dev->driver_data;
+	const struct dma_stm32_config *config = dev->config->config_info;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+	struct dma_stm32_stream *stream;
+	int id;
+
+	for (id = 0; id < data->max_streams; id++) {
+		if (func_is_active_tc[id](dma) ||
+		    func_is_active_ht[id](dma) ||
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+		    func_is_active_fe[id](dma) ||
+#endif
+		    func_is_active_te[id](dma)) {
+			break;
+		}
+	}
+
+	if (id == data->max_streams) {
+		LOG_ERR("Unknown interrupt happened.");
+		return;
+	}
+
+	stream = &data->streams[id];
+	stream->busy = false;
+	if (func_is_active_tc[id](dma)) {
+		func_clear_tc[id](dma);
+
+		stream->dma_callback(stream->callback_arg, id, 0);
+	} else if (func_is_active_ht[id](dma)) {
+		/* Silently ignore spurious transfer half complete IRQ */
+		LOG_INF("Half of the transfer has been completed.");
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
```

Comment:
Seems there are 2 different IPs to support in the driver.
In order to ease code reading, you can follow what has been done in SPI and I2C cases in which we declared Kconfig symbols that carry the IP variant information and then are selected by the series.
Cf  https://github.com/zephyrproject-rtos/zephyr/blob/528a2de29054c9234e19dd985f0fa8fb7dc05ee1/drivers/spi/Kconfig.stm32#L18
Youcan call them V1 and V2..
This way, when a new series require to support DMA, they only have to select the variant and we don't need to update the driver at each `#if define`

Reply:
Also, with this new symbol, you'll be able to use IS_ENABLED macro, which make things even easier to read.
You need to check how it works though.

Reply:
I can do this, but it may cost some time to finish it.